cache: stm32: add cortex-m33 peripheral driver

lindblandro opened the pull request:
STM32 Cortex-M33, such as the U5xxx series, have a cache peripheral for instruction and data caches, which are not present in the C-M33 architecture spec.

Closes #71268

vai-helin committed:
cache: stm32: add cortex-m33 peripheral driver

STM32 Cortex-M33, such as the L5/H5/U5 series, have a cache peripheral for
instruction and data caches, which are not present in the C-M33
architecture spec.

The driver defaults to direct mapped cache as it uses less power than the
alternative set associative mapping [1]. This has also been the default in
stm32 soc initialization code for chips that have the ICACHE peripheral,
which makes it the safest choice for backward compatibility. The exception
to the rule is STM32L5, which has the n-way cache mode selected in SOC
code.

[1]: https://en.wikipedia.org/wiki/Cache_placement_policies

Signed-off-by: Henrik Lindblom <henrik.lindblom@vaisala.com>

lindblandro commented:
Unsure where the cache driver should be enabled and which Kconfig options should be activated by default. This also now uses the STM32 HAL symbols (`DCACHE1` `DCACHE2`) directly in conditional compilation paths instead of checking for device support. Checking for chip type explicitly might be more readable.

I initially figured that the SOC Kconfig files would select `HAS_ICACHE` and `HAS_DCACHE` but that triggered a code path in the arch startup code that tried to use CMSIS to activate the Arm caches, which this core doesn't have.

JarmouniA reviewed (commented):
The following should be added to `Kconfig.defconfig` of M33 SoC series that have the cache.
```
config CACHE_STM32_CACHE
	default y if EXTERNAL_CACHE

config CACHE_MANAGEMENT
	default y

choice CACHE_TYPE
	default EXTERNAL_CACHE
endchoice
```
Also, all `LL_ICACHE_x` calls should be replaced by their corresponding `cache_instr_x` throughout the code.

lindblandro commented:
> The following should be added to `Kconfig.defconfig` of M33 SoC series that have the cache.
> 
> ```
> config CACHE_STM32_CACHE
> 	default y if EXTERNAL_CACHE
> 
> config CACHE_MANAGEMENT
> 	default y
> 
> choice CACHE_TYPE
> 	default EXTERNAL_CACHE
> endchoice
> ```
> 
> Also, all `LL_ICACHE_x` calls should be replaced by their corresponding `cache_instr_x` throughout the code.

Okay so I was initially in the right path as I had more or less what you suggest but I completely missed the `CACHE_TYPE` option, which explains why everything broke down. I'll add those in another commit. Thanks.

vai-helin committed:
soc: stm32: use cache peripheral driver

Use the Zephyr cache API in soc initialization code instead of calling the
HAL directly. The change does not modify the pre-existing cache settings,
just changes the path they are enabled.

Signed-off-by: Henrik Lindblom <henrik.lindblom@vaisala.com>

mathieuchopstm commented on drivers/cache/cache_stm32.c:
```diff
@@ -0,0 +1,162 @@
+
+int cache_data_flush_all(void)
+{
+	return -ENOTSUP;
```
Even if the hardware doesn't expose a "flush all" primitive, I think this should still be implemented as something like:
```c
return cache_data_manage_range(/* cache area base address */, /* cache area size */, CLEAN_BY_ADDR);
```

(if you look at the implemented of ARM's `SCB_InvalidateDCache()`, it works similarly)

mathieuchopstm commented on drivers/cache/cache_stm32.c:
```diff
@@ -0,0 +1,162 @@
+
+int cache_instr_flush_and_invd_all(void)
+{
+	int rc = cache_instr_flush_all();
```
This will return `-ENOTSUP`. Just make this routine return `-ENOTSUP` directly.

mathieuchopstm commented on drivers/cache/cache_stm32.c:
```diff
@@ -0,0 +1,162 @@
+
+int cache_instr_flush_and_invd_range(void *addr, size_t size)
+{
+	int rc = cache_instr_flush_range(addr, size);
```
Ditto. This will return `-ENOTSUP`, so just make this routine return `-ENOTSUP` directly.

mathieuchopstm commented on drivers/cache/cache_stm32.c:
```diff
@@ -0,0 +1,164 @@
+
+LOG_MODULE_REGISTER(cache_stm32, CONFIG_CACHE_LOG_LEVEL);
+
+static inline void wait_for_icache(void)
+{
+	while (LL_ICACHE_IsActiveFlag_BUSY()) {
+	}
+}
```
Nit: this has only one caller and is so short we should just inline it in `cache_instr_enable`

mathieuchopstm commented on drivers/cache/Kconfig.stm32:
```diff
@@ -0,0 +1,9 @@
+	select CACHE_HAS_DRIVER
+	depends on CACHE_MANAGEMENT
+	help
+	  Enable support for the STM32 ICACHE / DCACHE peripheral present in some Cortex-M33 chips.
```
Let's avoid this becoming stale with possible new products.
```suggestion
	  Enable support for the STM32 ICACHE / DCACHE peripheral present in some chips.
```

mathieuchopstm reviewed (commented):
Could you please modify the driver so that it works on SoCs without DCACHE? (e.g., STM32L5/STM32H503/STM32WBA)

(Assuming `CONFIG_DCACHE` is not defined on these series, I guess you could gate all `cache_data_*` functions behind it?)

erwango commented on drivers/cache/cache_stm32.c:
```diff
@@ -0,0 +1,164 @@
+
+void cache_instr_disable(void)
+{
+	LL_ICACHE_Disable();
```
In tree, we can find 2 alternatives for cache disabling:
```
	/* Disabling ICACHE */
	LL_ICACHE_Disable();
	while (LL_ICACHE_IsEnabled() == 1U) {
	}

	/* Wait until ICACHE_SR.BUSYF is cleared */
	while (LL_ICACHE_IsActiveFlag_BUSY() == 1U) {
	}

	/* Wait until ICACHE_SR.BSYENDF is set */
	while (LL_ICACHE_IsActiveFlag_BSYEND() == 0U) {
	}
```
and
```
	/* Clear BSYENDF flag first and then disable the instruction cache
	 * that starts a cache invalidation procedure
	 */
	CLEAR_BIT(ICACHE->FCR, ICACHE_FCR_CBSYENDF);

	LL_ICACHE_Disable();

	/* Get tick */
	tickstart = k_uptime_get_32();

	/* Wait for instruction cache to get disabled */
	while (LL_ICACHE_IsEnabled()) {
		if ((k_uptime_get_32() - tickstart) >
						ICACHE_DISABLE_TIMEOUT_VALUE) {
			/* New check to avoid false timeout detection in case
			 * of preemption.
			 */
			if (LL_ICACHE_IsEnabled()) {
				status = -ETIMEDOUT;
				break;
			}
		}
	}
```
We'll need to converge on one.

lindblandro replied:
The first method seems peculiar as it essentially boils down to this:

1. clear bit `ICACHE_CR_EN`
2. check if `ICACHE_CR_EN` is set
3. wait while cache is executing a CACHEINV operation indicated by `BUSYF`
4. wait for CACHEINV execution to finish indicated by `BSYENDF`

The second method seems also weird in that it does not wait for the cache invalidation operation but just checks if the `ICACHE_CR_EN` bit is set in a loop immediately after clearing it. Clearing the busy end flag might have something to do with interrupts, but the `IsEnabled()` still seems redundant. I might be missing something more complicated here though.

I'm not sure if there's anything that needs to be done after disabling the cache, which automatically triggers a full invalidation, but the cache is also bypassed for all accesses. The reference manual for STM32 H5, L5, U5, and WBA only mention

H5:
> Note: When disabled, the DCACHE is bypassed: slave input requests are just forwarded to the
master port.

L5, U5, and WBA:
> Note: When disabled, ICACHE is bypassed, except the remapping mechanism that is still
functional: the slave input requests (remapped or not) are just forwarded to the master
port(s).

So basically, you _might_ want to clear cache remap rules if disabling the cache, but that could also be really annoying if you just want to write something to flash and re-enabled the cache. Only time checking for ongoing invalidation is explicitly mentioned (that I noticed at least) is when enabling the cache.

lindblandro replied:
> > I'm not sure if there's anything that needs to be done after disabling the cache, which automatically triggers a full invalidation, but the cache is also bypassed for all accesses.
> 
> According to the ref implementation of cache disabling onnH5 by ST, there is. https://github.com/STMicroelectronics/stm32h5xx-hal-driver/blob/8c41bbe0b445f5036163d2ac637b50962d8b2f21/Src/stm32h5xx_hal_dcache.c#L414

That is the `DCACHE` where it's waiting currently executing management operations to finish, which is certainly something that should be done. I think ST's example code, the docs, and common sense align pretty well and I need to add a wait for the pending operations after disabling the cache.

Above I was referring to `ICACHE` which was the example presented. Now that you pointed the reference HAL drivers it's pretty clear where the second method of disabling the instruction cache comes from: https://github.com/STMicroelectronics/stm32u5xx-hal-driver/blob/main/Src/stm32u5xx_hal_icache.c#L239

Unfortunately it's impossible to say why ST has chosen to poll the enable bit for `ICACHE` but instead opts to check the busy flag for `DCACHE` (which makes more sense based on the docs). They have this piece of documentation in stmu5xxx-hal `ICACHE` driver :

```
        To ensure a deterministic cache behavior after power on, system reset or after
        a call to @ref HAL_ICACHE_Disable(), the application must call
        @ref HAL_ICACHE_WaitForInvalidateComplete(). Indeed on power on, system reset
        or cache disable, an automatic cache invalidation procedure is launched and the
        cache is bypassed until the operation completes.
```

which to me seems effectually the same as waiting after enabling (which is what the reference manuals suggest). For comparison, the exact quote from the [U5xxx reference manual RM0456](https://www.st.com/resource/en/reference_manual/rm0456-stm32u5-series-armbased-32bit-mcus-stmicroelectronics.pdf) is (§ 8.4.5):

> To ensure performance determinism, it is recommended to wait for the end of a potential
cache invalidate procedure before enabling the ICACHE.
> ...
>The software must test BUSYF and/or BSYENDF values before enabling the ICACHE.
Else, if the ICACHE is enabled before the end of an invalidate procedure, any cache access
(while BUSYF = 1) is treated as noncacheable, and its performance depends on the main
memory access time.

Could be that the reference is lacking and the actual behavior is that `EN` bit is not reset until `BUSYF` is cleared. 

I'm fine with waiting either before enabling or after disabling, but I'd like to understand why the `EN` bit is polled before implementing that.

Assuming that there is a valid reason for polling the enable status, would a spinlock be better than the somewhat cryptic double-checking stm32-hal drivers seem to do?

mathieuchopstm replied:
> > Note: When disabled, ICACHE is bypassed, except the remapping mechanism that is still
> > functional: the slave input requests (remapped or not) are just forwarded to the master
> > port(s).
> 
> So basically, you _might_ want to clear cache remap rules if disabling the cache, but that could also be really annoying if you just want to write something to flash and re-enabled the cache. The only instances explicitly mentioning checking for ongoing invalidation (that I noticed at least) is when enabling the cache.

I don't think we configure any such rule anyways, so I guess this can be ignored?

> Above I was referring to `ICACHE` which was the example presented. Now that you pointed the reference HAL drivers it's pretty clear where the second method of disabling the instruction cache comes from: https://github.com/STMicroelectronics/stm32u5xx-hal-driver/blob/main/Src/stm32u5xx_hal_icache.c#L239
> 
> Unfortunately it's impossible to say why ST has chosen to poll the enable bit for `ICACHE` but instead opts to check the busy flag for `DCACHE` (which makes more sense based on the docs). They have this piece of documentation in stmu5xxx-hal `ICACHE` driver :
> 
> ```
>         To ensure a deterministic cache behavior after power on, system reset or after
>         a call to @ref HAL_ICACHE_Disable(), the application must call
>         @ref HAL_ICACHE_WaitForInvalidateComplete(). Indeed on power on, system reset
>         or cache disable, an automatic cache invalidation procedure is launched and the
>         cache is bypassed until the operation completes.
> ```
> 
> which to me seems effectually the same as waiting after enabling (which is what the reference manuals suggest). For comparison, the exact quote from the [U5xxx reference manual RM0456](https://www.st.com/resource/en/reference_manual/rm0456-stm32u5-series-armbased-32bit-mcus-stmicroelectronics.pdf) is (§ 8.4.5):
> 
> > To ensure performance determinism, it is recommended to wait for the end of a potential
> > cache invalidate procedure before enabling the ICACHE.
> > ...
> > The software must test BUSYF and/or BSYENDF values before enabling the ICACHE.
> > Else, if the ICACHE is enabled before the end of an invalidate procedure, any cache access
> > (while BUSYF = 1) is treated as noncacheable, and its performance depends on the main
> > memory access time.
> 
> Could be that the reference is lacking and the actual behavior is that `EN` bit is not reset until `BUSYF` is cleared.
> 
> I'm fine with waiting either before enabling or after disabling, but I'd like to understand why the `EN` bit is polled before implementing that.
> 
> Assuming that there is a valid reason for polling the enable status, would a spinlock be better than the somewhat cryptic double-checking stm32-hal drivers seem to do?

From simple testing on U5 using the following:
```c
	LL_ICACHE_ClearFlag_BSYEND();
	printf("ENABLED=%d\n", LL_ICACHE_IsEnabled());             // 1
	printf("BUSY=%d\n", LL_ICACHE_IsActiveFlag_BUSY());        // 0
	printf("BSYENDF=%d\n", LL_ICACHE_IsActiveFlag_BSYEND());   // 0

	LL_ICACHE_Disable();
	printf("ENABLED=%d\n", LL_ICACHE_IsEnabled());             // 0

	while (LL_ICACHE_IsActiveFlag_BUSY()) {
            printf("BUSY=1\n"); //visible once if printf() after Disable() is removed
	}

	int busyend = LL_ICACHE_IsActiveFlag_BSYEND();
	printf("BUSY=%d\n", LL_ICACHE_IsActiveFlag_BUSY()); // 0
	printf("BSYENDF=%d\n", busyend);                    // 1
```

* `CR.EN` seems to be updated instantly (e.g., no need to poll it?)
  * The RM is not explicit about this but it makes sense
  * Cube implementation polling this bit is thus "wrong"/extra careful?
* `SR.BUSYF` is high while cache invalidation is ongoing
  * As documented, this operation starts right after setting `CR.EN=0`
* `SR.BSYENDF` goes high when `SR.BUSYF` goes low
  * If `BUSYF` is pollled, then polling `BSYENDF` is not necessary, as implied by RefMan: "_software must test BUSYF **and/or** BSYENDF_"

I think the following sequence would be correct:
```c
LL_ICACHE_Disable();

while (LL_ICACHE_IsActiveFlag_BUSY()) {
  /* Wait until ICACHE invalidation is performed */
}
```

But in my opinion, we don't even need to wait for `BUSYF` to go low during *disable*: the ICACHE is disabled as soon as `CR.EN=0` (which is seemingly instantaneous, though we could still poll it to be sure), and we wait for `BUSYF=0` before enabling anyways to follow RM recommendations.

mathieuchopstm replied:
TL;DR
```suggestion
	LL_ICACHE_Disable();
	
	while (LL_ICACHE_IsEnabled()) {
		/**
		 * Wait until the ICACHE is disabled (CR.EN=0), at which point
		 * all requests bypass the cache and are forwarded directly
		 * from the ICACHE slave port to the ICACHE master port(s).
		 *
		 * The cache invalidation will start once disabled, but we allow
		 * it to proceed in the background since it doesn't need to be
		 * complete for requests to bypass the ICACHE.
		 */
	 }
```

lindblandro replied:
Thanks for verifying this! I'll add the just-in-case check for `CR.EN` and update the PR.

lindblandro commented:
> Could you please modify the driver so that it works on SoCs without DCACHE? (e.g., STM32L5/STM32H503/STM32WBA)
> 
> (Assuming `CONFIG_DCACHE` is not defined on these series, I guess you could gate all `cache_data_*` functions behind it?)

Sure. I was already working on STM32WBA but wasn't aware of L5 / H503. The existing L5 soc init code doesn't set the 1-way cache like the others. Should that be changed in this PR or should the cache associativity be added as a configuration option?

The `cache_*` functions are already gated inside the `sys_cache_*` interface so they shouldn't end up in the final image. I guess they can be removed from the build completely.

mathieuchopstm commented:
> Sure. I was already working on STM32WBA but wasn't aware of L5 / H503. The existing L5 soc init code doesn't set the 1-way cache like the others. Should that be changed in this PR or should the cache associativity be added as a configuration option?

I guess it could be added as Kconfig option.

> The `cache_*` functions are already gated inside the `sys_cache_*` interface so they shouldn't end up in the final image. I guess they can be removed from the build completely.

The issue isn't about their presence in the binary, but rather that on such series, the `LL_DCache_xxx` functions most likely don't exist. Even if it's dead code, trying to use them from the driver should break the build if so.
(But if you're working on WBA, I guess you will see very quickly whether it builds without guards or not :slightly_smiling_face:)

lindblandro commented:
> > Sure. I was already working on STM32WBA but wasn't aware of L5 / H503. The existing L5 soc init code doesn't set the 1-way cache like the others. Should that be changed in this PR or should the cache associativity be added as a configuration option?
> 
> I guess it could be added as Kconfig option.
I wonder if there's any solid reasoning behind the current default for L5. The original commit didn't mention any and all the other chips use direct mapping as their default. I'll add the Kconfig option in any case.

> > The `cache_*` functions are already gated inside the `sys_cache_*` interface so they shouldn't end up in the final image. I guess they can be removed from the build completely.
> 
> The issue isn't about their presence in the binary, but rather that on such series, the `LL_DCache_xxx` functions most likely don't exist. Even if it's dead code, trying to use them from the driver should break the build if so. (But if you're working on WBA, I guess you will see very quickly whether it builds without guards or not 🙂)

Indeed I did :sweat_smile:.

vai-helin committed:
drivers: stm32: use cache peripheral driver

Use cache API for disabling and enabling ICACHE. The driver handles waiting
for ongoing cache invalidation.

Signed-off-by: Henrik Lindblom <henrik.lindblom@vaisala.com>

mathieuchopstm commented on soc/st/stm32/stm32u5x/power.c:
```diff
@@ -38,18 +38,7 @@ static void pwr_stop3_isr(const struct device *dev)
-	/* Wait until ICACHE_SR.BSYENDF is set */
-	while (LL_ICACHE_IsActiveFlag_BSYEND() == 0U) {
-	}
+	sys_cache_instr_disable();
```
Nit: since `disable_cache()` is now a mere proxy for `sys_cache_instr_disable()`, we should just get rid of it and call the `sys_*` function directly.

JarmouniA reviewed (changes requested):
Issue link to be removed from commit messages https://docs.zephyrproject.org/latest/contribute/guidelines.html#adding-links

lindblandro commented on drivers/flash/flash_stm32l5x.c:
```diff
@@ -302,19 +222,15 @@ int flash_stm32_block_erase_loop(const struct device *dev,
+	 */
+
+	LOG_DBG("I-cache Disable");
+	sys_cache_instr_disable();
```
I think the API could have a method for checking if the cache is enabled to hide the details from drivers. Using the `sys_cache_` API might allow combining the various stm32 flash drivers for example. It's of course possible with a few `LL_ICACHE_*` calls sprinkled here and there but the code would be cleaner if the cache API allowed checking cache status.

In any case, I'll re-add the conditional enable logic using the LL API.

mathieuchopstm replied:
Indeed, a `sys_cache_X_is_enabled()` API would be useful. Definitely outside this PR's scope though, so LL API it is :slightly_smiling_face:

lindblandro commented on drivers/sensor/st/stm32_vref/stm32_vref.c:
```diff
@@ -103,14 +103,14 @@ static int stm32_vref_channel_get(const struct device *dev, enum sensor_channel
  */
 #if defined(CONFIG_SOC_SERIES_STM32H5X)
-	LL_ICACHE_Disable();
+	sys_cache_instr_disable();
```
Yes I think `EXTERNAL_CACHE` and `ICACHE` could be used. This explanation by @mathieuchopstm https://github.com/zephyrproject-rtos/zephyr/pull/57178#discussion_r1677374161 points out the issue with cache refills. The issue is with cache line width, which the application note [AN5212](https://www.st.com/resource/en/application_note/an5212-how-to-use-stm32-cache-to-optimize-performance-and-power-efficiency-for-stm32-mcus-stmicroelectronics.pdf)  summarizes as 128bits for all mentioned devices. The reference manuals for U5 is a bit coy on the matter, but it does mention that OTP is through register reads only and the cache line width is naturally documented without mention to these parts of the flash. 

However, the cache implementation might be better and the OTP area could in theory be treated as non-cacheable. I'm mentioning this as I've been using both vref and temp sensors in U585 and don't recall any bus errors even though ICACHE was enabled by soc init code. I'll run a quick check tomorrow and compare H5 and U5 RMs a bit more thoroughly.

The docs could definitely be a lot better on this matter.

lindblandro replied:
I tested simply with
```
 #if defined(CONFIG_SOC_SERIES_STM32H5X)
 	sys_cache_instr_disable();
 #endif /* CONFIG_SOC_SERIES_STM32H5X */
-
+	LOG_INF("ICACHE is: %s", LL_ICACHE_IsEnabled() ? "enabled" : "disabled");
 	/* Calculate VREF+ using VREFINT bandgap voltage and calibration data */
 	vref = (cfg->cal_mv * ((*cfg->cal_addr) >> cfg->cal_shift)) / data->raw;
```
and I get this in the logs:
```
[00:01:34.086,000] <inf> stm32_vref: ICACHE is: enabled
```
while reading vref (and core temperature) in a loop. This is with board `b_u585i_iot02a/stm32u585xx/ns`. So it seems it's working better at least in U585.

mathieuchopstm replied:
> The reference manuals for U5 is a bit coy on the matter, but it does mention that OTP is through register reads only and the cache line width is naturally documented without mention to these parts of the flash.

Can you point out which part of the RM? I can't seem to find such a notice :thinking:

lindblandro replied:
> > The reference manuals for U5 is a bit coy on the matter, but it does mention that OTP is through register reads only and the cache line width is naturally documented without mention to these parts of the flash.
> 
> Can you point out which part of the RM? I can't seem to find such a notice 🤔

Hmm, it was getting late and I might've mixed OTP and option bytes while reading section § 7.3.1 where it says

```
- 512 bytes OTP (one-time programmable) bytes for user data (32 quad-words).
The OTP data cannot be erased and can be written only once.
- option bytes for user configuration. Unlike user flash memory and system memory,
it is not mapped to any memory address and can be accessed only through the
flash register interface.
```

I think I just blasted through those two paragraphs, ignoring the second bullet where it starts to talk about option bytes. Sorry for the confusion.

The VREF calibration register is located after the OTP block at `0x0BFA07A5` (OTP ends at `0x0BFA0200`). That and the internal temperature sensor registers seem to work with ICACHE in any case.

erwango commented on drivers/cache/Kconfig.stm32:
```diff
@@ -0,0 +1,24 @@
+
+if CACHE_STM32
+
+# "default n" for L5 is legacy - could be removed?
+config CACHE_STM32_ICACHE_DIRECT_MAPPING
+	bool "Use 1-way associative mapping for ICACHE"
+	default n if SOC_SERIES_STM32L5X
```
I think it should be ok to move L5 to 1-way associated mapping, but we should document the change in migration guide and mention how to revert.
